reconcile/unleash_feature_toggles/integration.py

```diff
@@ -0,0 +1,310 @@
+                feature_type = FeatureToggleType[add.unleash.q_type]
+            except KeyError:
+                raise ValueError(
+                    f"[{instance.name}/{project_id}/{add.name}] Invalid feature toggle type '{add.unleash.q_type}', Possible values are: {', '.join(FeatureToggleType.__members__)}"
+                )
+            if not dry_run:
+                client.create_feature_toggle(
+                    project_id=project_id,
+                    name=add.name,
+                    description=add.description,
+                    type=feature_type,
+                    impression_data=bool(add.unleash.impression_data),
+                )
+
+        for change in diff.change.values():
+            logging.info(
+                f"[{instance.name}/{project_id}] Changing feature toggle {change.desired.name}"
+            )
+            try:
+                assert change.desired.unleash.q_type  # make mypy happy
+                feature_type = FeatureToggleType[change.desired.unleash.q_type]
+            except KeyError:
+                raise ValueError(
+                    f"[{instance.name}/{project_id}/{change.current.name}] Invalid feature toggle type '{change.desired.unleash.q_type}', Possible values are: {', '.join(FeatureToggleType.__members__)}"
+                )
+            if not dry_run:
+                client.update_feature_toggle(
+                    project_id=project_id,
+                    name=change.current.name,
+                    description=change.desired.description,
+                    type=feature_type,
+                    impression_data=bool(change.desired.unleash.impression_data),
+                )
+
+        for delete in diff.delete.values():
+            desired_toggle = next(
+                (d for d in desired_state if d.name == delete.name and d.delete), None
+            )
+            if desired_toggle:
+                logging.info(
+                    f"[{instance.name}/{project_id}] Deleting feature toggle {delete.name}"
+                )
+                if not dry_run:
+                    client.delete_feature_toggle(
+                        project_id=project_id,
+                        name=delete.name,
+                    )
+            elif not instance.allow_unmanaged_feature_toggles:
+                raise UnleashFeatureToggleDeleteError(
+                    f"[{instance.name}/{project_id}] Found unmanaged feature toggles '{[d.name for d in diff.delete.values()]}'"
+                )
+
+    def _reconcile_states(
+        self,
+        client: UnleashServer,
+        instance: UnleashInstanceV1,
+        project_id: str,
+        dry_run: bool,
+        current_state: Sequence[FeatureToggle],
+        desired_state: Iterable[FeatureToggleUnleashV1],
+        available_environments: Iterable[Environment],
+    ) -> None:
+        """Reconcile the feature toggle states."""
+        # Manage the actual feature toggle states
```

Comment:
I thought we don't manage toggle state, or this is something else?

Reply:
You can now manage the state as well, but it's optional.

Reply:
so when it's defined like 

```yaml
 environments:
    default: true
```

the integration will reset it to `on` if toggle off manually in unleash.

Another question is whether enable or disable project `environment` itself can affect toggle state?

Reply:
<img width="1217" alt="Screenshot 2024-05-16 at 4 16 23 PM" src="https://github.com/app-sre/qontract-reconcile/assets/982709/aee816ea-4350-4748-acb9-8a29fc835564">
will this affect fetched toggle state value?

Reply:
btw to simplify state management, we can flatten all toggles and envs to make diff easier, like this:

```python
        current_toggle_states = {
            (state.name, env.name): env.enabled
            for state in current_state
            for env in state.environments
        }
        desired_toggle_states = {
            (state.name, env_name): enabled
            for state in desired_state
            for env_name, enabled in (state.unleash.environments or {}).items()
        }

        diff_result = diff_mappings(current=current_toggle_states, desired=desired_toggle_states)
        if diff_result.add:
            raise ValueError("missing xxxx")
        for (name, env), pair in diff_result.change.items():
            if not dry_run:
                client.set_feature_toggle_state(
                    project_id=project_id,
                    name=name,
                    environment=env,
                    enabled=pair.desired,
                )
```

Reply:
No, disabled projects are ignored

Reply:
how is that ignored work? when fetch `features`, returned `environments` will auto exclude disabled ones?

Reply:
> how is that ignored work? when fetch features, returned environments will auto exclude disabled ones?

Exactly, disabled environments are not listed in the `/api/admin/projects/{project_id}/features` results.

Reply:
> btw to simplify state management, ....

change in the latest commit